src/core/ipns/resolver.js

```diff
@@ -0,0 +1,66 @@
+'use strict'
+
+const IpnsEntry = require('./pb/ipnsEntry')
+const utils = require('./utils')
+const validator = require('./validator')
+
+const multihash = require('multihashes')
+
+class IpnsResolver {
+  constructor (repo) {
+    this.repo = repo
+  }
+
+  resolve (name, publicKey, options, callback) {
+    const nameSegments = name.split('/')
+
+    if (nameSegments.length !== 3 || nameSegments[0] !== '') {
+      return callback(new Error(`invalid name syntax for ${name}`))
+    }
+
+    const key = nameSegments[2]
+
+    // TODO recursive
+    // TODO nocache
+
+    if (options.local) {
+      this.resolveLocal(key, publicKey, (err, res) => {
+        if (err) {
+          return callback(err)
+        }
+
+        return callback(null, res)
+      })
+    } else {
+      return callback(new Error('not implemented yet'))
+    }
+  }
+
+  // https://github.com/ipfs/go-ipfs-routing/blob/master/offline/offline.go
+  resolveLocal (name, publicKey, callback) {
```

Comment:
Eventually, this should be pushed into a non-IPNS specific routing layer.

Reply:
I had been thinking about the `go-ipfs` implementation for the resolver. 

For one side, I agree with your [approach](https://github.com/ipfs/go-ipfs/blob/master/core/commands/ipns.go#L93). That is, you decide which router the IPNS resolver should use, and this router is completely transparent for the IPNS resolver. However, as we are fetching from an offline and local source (datastore), it also seems strange to me to have the `OfflineRouter`. 

Anyway, I am not sure about where this code should live on. @diasdavid any opinion here?

Reply:
Basically, it means that we can swap out the router at will. However, given the fact that you're only implementing offline routing, it doesn't really make a difference.

Reply:
Being a pluggable router is a considerable advantage. But yeah, that's what I thought while putting the `resolveLocal` in there. 

However, when I integrate the routing I will have to understand if it still makes sense to have the `resolveLocal` inside the IPNS layer.